### ISSUE TITLE
Reconsider strategy for implicitly generating NN derivatives

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#3874]
At the moment, the NN story looks something like this:

```
# LOOKING FOR AN THNN DEFINITION?  These are implicitly defined by
# preprocess_nn_functions, because THNN functions follow a specific
# convention; thus, you don't have to explicitly write it down in
# derivatives.yaml (in fact, you MUST NOT write it down, because then
# you'll have two derivative definitions!)
```

As I have been working on CuDNN bindings, I've found that the implicit generation of NN derivatives has been getting in my way more than it has been helping. The two big things I've noticed:

* It's easy to "forget" that an NN definition is being implicitly generated, and accidentally specify its derivative redundantly in `derivatives.yaml`. Right now the generator code silently accepts this and generates two classes in this case.
* It's easy to accidentally trigger the NN derivative generation behavior when you don't want to; it's sufficient to have two functions `foo` and `foo_backward` and blam, NN definitions get generated. But the native function parsing machinery is not setup to handle this yet, since the NN autogen code assumes outputs are named, but the native parser doesn't exercise this at all. Also, you need to define BOTH `foo` and `foo_forward` for the codepath to work. In short, there are some pretty stringent assumptions about what "NN" things look like, which are not very friendly to native function writers.

So you end up with things like RoiPooling "near-missing" the NN autogen behavior by naming their forward function `RoiPooling2D_forward` (and thus getting a backwards class named `RoiPooling2DForward`), avoiding the NN codepath by a hair.

This just seems bad.

Here are two possible proposals:

* Remove NN autogen behavior. We explicitly write down derivatives for all the NN classes; none of this magic stuff. Codegen gets simpler.

* Make the NN autogeneration behavior more widely applicable. For example, today, for the bindings to work you must define both `foo`, `foo_forward`, and `foo_backward`; if you're missing any of these the tooling chokes. But this is fairly unfriendly for native functions.

**Comment by colesbury** [comment]
Yes please, let's do (1) "Remove NN autogen behavior"

**Comment by lantiga** [comment]
I'm taking care of this

**Comment by lantiga** [comment]
Working on it 😅

**Comment by lantiga** [comment]
@ezyang I’m going ahead with this in my branch.
I’ve generated a crwap file for nn. One doubt: do we also want to migrate from cwrap to yaml proper as part of this effort?

**Comment by colesbury** [comment]
We don't want any more cwrap files. The derivative definitions should be in `derivatives.yaml`. The rest of the NN binding is fine for now.

This is the part which should be replaced by explicit definitions in `derivatives.yaml`:

https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L453-L523

**Comment by lantiga** [comment]
I should have provided more context: I had a chat with @ezyang and we talked about also getting rid of parse_nn.py in aten/src/ATen. This is where a cwrap declaration file for nn functions came about.

I was wondering if we could further simplify things by migrating to yaml proper for all declaration files. In any case let me know if I should stop here and leave parse_nn.py in place.

Branch is here, just for reference: https://github.com/lantiga/pytorch/tree/no_autogen

**Comment by colesbury** [comment]
I don't think that's a worthwhile effort. I don't think moving the NN binding code to cwrap makes sense.

For future stuff functions and layers, we should implement them direclty in ATen.

**Comment by lantiga** [comment]
👍 Sounds good, thanks for the quick feedback. I’ll complete the work on derivatives and submit a PR

**Comment by ezyang** [comment]
OK, reading @colesbury's comments, it makes sense that we're not going to touch the actual THNN/THCUNN bindings for now. Sorry about the confusion!

**Comment by lantiga** [comment]
@ezyang can’t we just change names in aten/src/ATen/nn.yaml ?

**Comment by colesbury** [comment]
Yeah, I was expected we'd rename batch norm and conv in `nn.yaml` (to something like `_thnn_batch_norm`) and have the public `at::batch_norm` do the correct dispatching.

**Comment by lantiga** [comment]
I guess we can close this

**Comment by zou3519** [comment]
This issue seems out of date so I am closing it. Please feel free to reopen if that is not the case.
